### ISSUE TITLE
⚗️ Proposal for checking service status in the `dynamic-scheduler`

### ISSUE THREAD

**Issue by GitHK** [ITISFoundation/osparc-simcore#5347]
# Introduction

## `Problem` in short

The `dynamic-scheduler` needs to monitor the status of all the dynamic services
that are currently tracked by it.

##### What is it used for?

A service responsible for keeping track dynamic services.
It's main job is to keep track and enforce the state of the services.

##### Design philosophy

- The `dynamic-scheduler` is always ran in multiple copies `2+`.
- The service is allowed to be restarted without warning at any time. Meaning,
  by design the service is allowed to fail and should recover it's work
  form it's previous point.
- If another copy of the `dynamic-scheduler` is running and one of them dies
  the running copy should take over the jobs of the previous one.
- The `dynamic-scheduler` can restart at any given time.
- There is no guarantee about the number of running copies.
- If third party software like (RabbitMQ and Redis) become unavailable
  the service because unhealthy and is restarted.

### Detailed request

- The `dynamic-scheduler` can keep track of a very large amount of dynamic services.
  Avoid overwhelming the system with requests all at the same time.
  (example: avoid requesting the status of all services at the exact same instant)
- Running the `dynamic-scheduler` in one or multiple copies should not have an impact.
- Scaling the `dynamic-scheduler` must not have an impact.
- Randomly killing all running copies, except one, should not have an impact.
- Depending on the current state of the service, the `dynamic-scheduler` can modify the interval
  at which it fetches status updates. Currently only the following intervals are planned
  to be used: `1s`, `5s` and `60s`.

---

# Outcomes and Special Scenarios

## Desired `outcome`

As a developer, I need to schedule actions at a desired time in the future.
A simplified version of this is enough: schedule an action every "x" seconds.

To achieve this:
- As a developer I want to be able to schedule an `async handler` to run at an established interval.
- As a developer I need to change the interval of the registered `async handler` without
  removing and rescheduling it.
- At any moment it should be possible to remove the an `async handler`. This also implies cancelling
  future scheduled runs.
- The `async handler` should only run only once, even if there are multiple copies of the
  `dynamic-scheduler` running.
- If an `async handler` fails to execute (raises an error), log the error and do not retry the execution.
- If the run of the previous `async handler` call is still ongoing (taking longer than expected) the current
  action should be skipped, note that the next event should be queued to be executed.

## List of `edge cases`

- If there are 0 copies of the `dynamic-scheduler` it is ok to not have any status updates.

---

# Proposal

## `Solution` in short

A distributed scheduler is what is required here. To build one with all the above constraints in mind the following
are needed:
1. A task that runs at a fixed interval (only once across all instances) and enforces the scheduled `async handler` configurations
1. Distributed shared memory (`Redis`) were we keep track of the current `async handler` schedule.
1. A way to execute an event in the future.
1. A worker pool to take incoming `async handler` execution requests.

### 1. Task that scheduler configurations

- a task thar runs only once every "y" second, across all instances, and checks that the schedule configurations are still valid (checks if there were any missed `async handler` invocations)
- if missed invocations are detected it starts the scheduling cycle again

### 2. Distributed shared memory
- in Redis store the current state of the `async handler` schedule
- the state contains the following:
  - `last invocation`: if current time - next invocation interval < than last invocation =>  it did not schedule a new task to be invoked (recreate task)
  - `current invocation counter`: used to figure out if the correct schedule event was invoked; each time the event is scheduled this counter is attached, if it differs from what is expected, it discards the event
  - `nex invocation interval`: amount by which th next invocation of `async handler` should be delayed

### 3. Schedule event in the future
- Best option for distributed systems, that I am aware of, is using a message queue with delayed delivery (implemented with `RabbitMQ` and `dead letter exchange` pattern).
- For current requirements, 4 different queues are enough:
  - 0s delay: deliver immediately, used to execute action immediately (when handler is first scheduled)
  - 1s, 5s and 60s delay: intervals at which the `dynamic-scheduler` decides to poll the service status (when the service is starting, stopping and running, different intervals are required)

### 4. Worker pool
A similar concept to the ThreadPoolExecutor will be used for `async handler`.
- can enqueue new calls at any given moment (gather requires all the coroutines to be created first)
- allows to limit parallelism
- dequeue events from RabbitMQ as fast as possible and push them to the pool
- dequeuing 1 job at a time, should allow for equal distribution between all instances of the `dynamic-scheduler`

### Expected behaviour

- All instances of `dynamic-scheduler` working: everything works as expected, the instances share the load and `async handler` invocation should be distributed between them
- If 1 instance of the `dynamic-scheduler` reboots: expect no issues to occur
- When all instances of `dynamic-scheduler` reboot: it could happen that some registered `async handler` future calls will be emitted. The schedules will have to be recreated.

**Comment by sanderegg** [comment]
are you implementing [Celery](https://docs.celeryq.dev/en/stable/getting-started/introduction.html)?

**Comment by GitHK** [comment]
> are you implementing [Celery](https://docs.celeryq.dev/en/stable/getting-started/introduction.html)?
Not at all.

I have evaluated `celery` and `scheduler`. None were sufficient for this application.
Unfortunately I have to build a scheduler.

**Comment by bisgaard-itis** [comment]
> > are you implementing [Celery](https://docs.celeryq.dev/en/stable/getting-started/introduction.html)?
> 
> Not at all.
> 
> I have evaluated `celery` and `scheduler`. None were sufficient for this application. I have to build a lightweight scheduler with the requirements mentioned above.

Very interesting! But it also looks like quite a mouthful! Would it not be possible to build this on top of Celery or some other well supported tool?

**Comment by sanderegg** [comment]
I do agree with @bisgaard-itis on this. if possible using an already well tested tool and build on top is a better option I think than creating everything from scratch.

**Comment by GitHK** [comment]
@bisgaard-itis @sanderegg 
I would like to use a pre baked solution.

My main difficulty here is that, while the service is running I need to do the following:
- schedule a new async handler
- modify the schedule period of an existing async handler
- remove and an async handler

With `celery` is a bit complex to do so. It's a bit more statically configured.
I will have to implement the trie logic for scheduling and the only advantage I would get is run an event after "x" seconds done by celery. I can already have that with `RabbitMQ` with much, much less complexity.
And as a final argument against `celery`, it does not play nice with `async`.

Any other suggestions that I can investigate are welcomed.

**Comment by matusdrobuliak66** [comment]
> @bisgaard-itis @sanderegg I would like to use a pre baked solution.
> 
> My main difficulty here is that, while the service is running I need to do the following:
> 
> * schedule a new async handler
> * modify the schedule period of an existing async handler
> * remove and an async handler
> 
> With `celery` is a bit complex to do so. It's a bit more statically configured. I will have to implement the trie logic for scheduling and the only advantage I would get is run an event after "x" seconds done by celery. I can already have that with `RabbitMQ` with much, much less complexity. And as a final argument against `celery`, it does not play nice with `async`.
> 
> Any other suggestions that I can investigate are welcomed.

Maybe check https://pypi.org/project/aio-celery/ whether it would be useful?

**Comment by GitHK** [comment]
> > @bisgaard-itis @sanderegg I would like to use a pre baked solution.
> > My main difficulty here is that, while the service is running I need to do the following:
> > 
> > * schedule a new async handler
> > * modify the schedule period of an existing async handler
> > * remove and an async handler
> > 
> > With `celery` is a bit complex to do so. It's a bit more statically configured. I will have to implement the trie logic for scheduling and the only advantage I would get is run an event after "x" seconds done by celery. I can already have that with `RabbitMQ` with much, much less complexity. And as a final argument against `celery`, it does not play nice with `async`.
> > Any other suggestions that I can investigate are welcomed.
> 
> Maybe check https://pypi.org/project/aio-celery/ whether it would be useful?

Being able to use async is just a minor detail.
I already said that I have discarded `celery`. It is not suited for this application. I'm searching for something that can fit to my requirements and `celery` is not it.

**Comment by GitHK** [comment]
this is already under way closing issue.
